File: android/app/src/main/java/org/securechat/app/RelayUrls.kt
package org.securechat.app

import android.net.Uri

/**
 * The relay location the app talks to, split into the exact origins the client
 * and the CSP need. Parsing is strict: an http(s) scheme, a host, and no path —
 * a relay is an origin, not a URL into one.
 */
data class RelayUrls(
    val httpOrigin: String, // e.g. https://relay.example.com  (used for /api fetch)
    val wsOrigin: String,   // e.g. wss://relay.example.com    (client appends /ws)
) {
    companion object {
        // A relay host may only contain characters that are unambiguous in a
        // hostname. android.net.Uri does NOT validate host characters — it
        // happily returns `"`, `;`, `{`, `}`, spaces, and percent-decoded
        // delimiters in .host. Because the parsed value is later interpolated
        // into injected JS and the CSP header, a permissive host is an arbitrary
        // code-execution vector (see the 2026-07-08 Android pentest, Finding 1).
        // Whitelist letters, digits, dot and hyphen; that covers DNS names and
        // IPv4/.onion, and rejects IPv6 literals ("[...]") — unneeded here.
        private val SAFE_HOST = Regex("^[A-Za-z0-9.-]+$")

        fun parse(raw: String): RelayUrls? {
            val text = raw.trim().trimEnd('/')
            if (text.isEmpty()) return null
            val uri = runCatching { Uri.parse(text) }.getOrNull() ?: return null
            val scheme = uri.scheme?.lowercase() ?: return null
            if (scheme != "http" && scheme != "https") return null
            val host = uri.host ?: return null
            if (host.isEmpty()) return null
            // Fail closed on any host character outside the safe set. This is the
            // primary defense: it stops the parsed value from ever carrying quote,
            // brace, semicolon, or delimiter characters downstream.
            if (!SAFE_HOST.matches(host)) return null
            // Reject credentials in the authority (user:pass@host).
            if (uri.userInfo != null) return null
            // Reject anything carrying a path/query/fragment: origins only.
            if (!uri.path.isNullOrEmpty() || uri.query != null || uri.fragment != null) return null

            val portPart = if (uri.port != -1) ":${uri.port}" else ""
            val wsScheme = if (scheme == "https") "wss" else "ws"
            return RelayUrls(
                httpOrigin = "$scheme://$host$portPart",
                wsOrigin = "$wsScheme://$host$portPart",
            )
        }
    }
}

File: ios/SecureChat/RelayUrls.swift
import Foundation

/// The relay the app talks to, split into the exact origins the client and the
/// CSP need. Port of `android/.../RelayUrls.kt`, with one iOS-specific rule:
/// plain `http` is accepted for a LOOPBACK host only.
///
/// Why stricter than Android: the page is a secure context, so WebKit's
/// mixed-content blocker refuses `ws://` to anything but loopback, and App
/// Transport Security refuses cleartext below that (Info.plist allows local
/// networking only). A LAN or `.onion` http relay therefore could never
/// connect; refusing it at input time is a clear error instead of a silent
/// dead app. (There is no Tor on iOS in this app, so `.onion` is out either
/// way — see ios/README.md.)
struct RelayUrls: Equatable {
    let httpOrigin: String  // e.g. https://relay.example.com (fetch /api)
    let wsOrigin: String    // e.g. wss://relay.example.com   (client appends /ws)

    /// Letters, digits, dot and hyphen. The parsed value is interpolated into
    /// the injected script and the CSP, so a permissive host is a code-execution
    /// vector (2026-07-08 Android pentest, Finding 1). Checked on the
    /// PERCENT-ENCODED host, so an encoded delimiter fails here too.
    private static let safeHost = CharacterSet(charactersIn:
        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.-")

    private static let loopbackHosts: Set<String> = ["127.0.0.1", "localhost"]

    static func parse(_ raw: String) -> RelayUrls? {
        var text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        while text.hasSuffix("/") { text.removeLast() }
        guard !text.isEmpty, text.count <= 300 else { return nil }
        guard let c = URLComponents(string: text) else { return nil }
        guard let scheme = c.scheme?.lowercased(), scheme == "http" || scheme == "https" else { return nil }
        guard let host = c.percentEncodedHost, !host.isEmpty,
              host.unicodeScalars.allSatisfy({ safeHost.contains($0) }) else { return nil }
        // No credentials, and an origin only: no path, query or fragment.
        guard c.percentEncodedUser == nil, c.percentEncodedPassword == nil else { return nil }
        guard c.percentEncodedPath.isEmpty, c.percentEncodedQuery == nil,
              c.percentEncodedFragment == nil else { return nil }
        if scheme == "http" && !loopbackHosts.contains(host.lowercased()) { return nil }
        var portPart = ""
        if let port = c.port {
            guard (1...65535).contains(port) else { return nil }
            portPart = ":\(port)"
        }
        let ws = scheme == "https" ? "wss" : "ws"
        return RelayUrls(httpOrigin: "\(scheme)://\(host)\(portPart)",
                         wsOrigin: "\(ws)://\(host)\(portPart)")
    }
}
